app/coffee/RateLimitManager.coffee

```diff
@@ -0,0 +1,36 @@
+Settings = require('settings-sharelatex')
+logger = require('logger-sharelatex')
+Metrics = require('./Metrics')
+
+module.exports = class RateLimiter
+
+	constructor: (number = 10) ->
+		@ActiveWorkerCount = 0
+		@CurrentWorkerLimit = number
+		@BaseWorkerCount = number
+
+	_adjustLimitUp: () ->
+		@CurrentWorkerLimit += 0.1 # allow target worker limit to increase gradually
+
+	_adjustLimitDown: () ->
+		@CurrentWorkerLimit = Math.max @BaseWorkerCount, (@CurrentWorkerLimit * 0.9)
+		logger.log {currentLimit: Math.ceil(@CurrentWorkerLimit)}, "reducing rate limit"
+
+	_trackAndRun: (task, callback = () ->) ->
+		@ActiveWorkerCount++
+		Metrics.gauge "processingUpdates", "+1"  # increments/decrements gauge with +/- sign
+		task (err) =>
+			@ActiveWorkerCount--
+			Metrics.gauge "processingUpdates", "-1"
+			callback(err)
```

Comment:
Will tasks ever need more than an `err/null` in their callback functions?

It seems like values could be silently lost this way.